File: LowestCommonAncestor/lca.py
# Lowest Common Ancestor solution in Python 2 for CodeEval.com by Steven A. Dunn

class Node:
  def __init__(self):
    left = None
    right = None
    val = 0

def createTree():
  node = Node()
  node.val = 30

  node.left = Node()
  node.left.val = 8
  node.left.left = Node()
  node.left.left.val = 3
  node.left.left.left = None
  node.left.left.left = None
  node.left.right = Node()
  node.left.right.val = 20
  node.left.right.left = Node()
  node.left.right.left.val = 10
  node.left.right.left.left = None
  node.left.right.left.right = None
  node.left.right.right = Node()
  node.left.right.right.left = None
  node.left.right.right.right = None
  node.right = Node()
  node.right.val = 52
  node.right.left = None
  node.right.right = None

  return node

def findLCA(node, left, right):
  val = node.val
  if left < val and right > val: return val
  elif left > val: return findLCA(node.right, left, right)
  else: return findLCA(node.left, left, right)

import sys

for line in open(sys.argv[1], 'r'):
  line = map(int, line.rstrip('\n').split(' '))
  node = createTree()
  print findLCA(node, line[0], line[1])

File: LowestCommonAncestor/cpp/lca.cpp
// Lowest Common Ancestor solution in C++ for CodeEval.com by Steven A. Dunn

#include <cstdlib>
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

using std::cout;
using std::endl;
using std::ifstream;
using std::string;
using std::vector;

typedef struct Node
{
	Node* left;
	Node* right;
	int val;
} Node;

Node* createTree();
int findLCA(Node*, int, int);

int main(int argc, char* argv[])
{
	ifstream inputFile(argv[1]);
	string line;

	if (inputFile)
	{
		Node* node = createTree();
		while (getline(inputFile, line))
		{
			int spacePos = line.find(" ");
			vector<int> nodeVals(2, 0);
			nodeVals[0] = atoi(line.substr(0, spacePos).c_str());
			nodeVals[1] = atoi(line.substr(spacePos).c_str());

			int left = nodeVals[0];
			int right = nodeVals[1];

			if (left > right)
			{
				int tmp = left;
				left = right;
				right = tmp;
			}

			int LCA = findLCA(node, left, right);
			cout << LCA << endl;
		}
		inputFile.close();
	}
	return 0;
}

Node* createTree()
{
	Node* node = new Node();
	node->val = 30;

	node->left = new Node();
		node->left->val = 8;
		node->left->left = new Node();
			node->left->left->val = 3;
			node->left->left->left = NULL;
			node->left->left->left = NULL;
		node->left->right = new Node();
			node->left->right->val = 20;
			node->left->right->left = new Node();
			node->left->right->left->val = 10;
				node->left->right->left->left = NULL;
				node->left->right->left->right = NULL;
			node->left->right->right = new Node();
			node->left->right->right->val = 29;
				node->left->right->right->left = NULL;
				node->left->right->right->right = NULL;
	node->right = new Node();
		node->right->val = 52;
		node->right->left = NULL;
		node->right->right = NULL;

	return node;
}

int findLCA(Node* node, int val1, int val2)
{
	int nodeVal = node->val;
	if (nodeVal == val1 || nodeVal == val2)
		return nodeVal;
	if (val1 < nodeVal && val2 > nodeVal)
		return nodeVal;
	else if (val1 > nodeVal)
		return findLCA(node->right, val1, val2);
	else
		return findLCA(node->left, val1, val2);
}